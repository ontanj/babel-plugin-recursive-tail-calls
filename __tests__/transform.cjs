const babel = require('@babel/core');
const plugin = require("..");

const base = `
function base(a) {
  if (a <= 0) {
    return a;
  }
  return base(a - 1);
}
`;

const multipleArgs = `
function multipleArgs(a, b) {
  if (a < b) return a
  return multipleArgs(a - 1, b + 1)
}
`;

const defaultArgInactive = `
function defaultArgInactive(a, b = 0) {
  if (a < b) return a
  return defaultArgInactive(a - 1, b + 1)
}
`;

const defaultArgActive = `
function defaultArgActive(a, b = 0) {
  if (a < b) return a
  return defaultArgActive(a - 1)
}
`;

const undefinedArg = `
function undefinedArg(a, b) {
  if (a < b) return a
  return undefinedArg(a - 1)
}
`;

const arrowFunction = `
const arrowFunction = (a) => {
  if (a <= 0) return a;
  return arrowFunction(a - 1)
}
`;

const cases = Object.entries({
  base,
  multipleArgs,
  defaultArgInactive,
  defaultArgActive,
  undefinedArg,
  arrowFunction
})

describe("transform", () => {
  it.each(cases)('%s', (_, testCase) => {
    const { code } = babel.transform(testCase, { plugins: [plugin] });
    expect(code).toMatchSnapshot();
  });
});

