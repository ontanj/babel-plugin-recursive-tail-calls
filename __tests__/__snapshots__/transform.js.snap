// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform arrowFunction 1`] = `
"const arrowFunction = a => {
  while (true) {
    if (a <= 0) return a;
    a = a - 1;
  }
};"
`;

exports[`transform base 1`] = `
"function base(a) {
  while (true) {
    if (a <= 0) {
      return a;
    }
    a = a - 1;
  }
}"
`;

exports[`transform defaultArgActive 1`] = `
"function defaultArgActive(a, b = 0) {
  while (true) {
    if (a < b) return a;
    a = a - 1;
    b = 0;
  }
}"
`;

exports[`transform defaultArgInactive 1`] = `
"function defaultArgInactive(a, b = 0) {
  while (true) {
    if (a < b) return a;
    a = a - 1;
    b = b + 1;
  }
}"
`;

exports[`transform multipleArgs 1`] = `
"function multipleArgs(a, b) {
  while (true) {
    if (a < b) return a;
    a = a - 1;
    b = b + 1;
  }
}"
`;

exports[`transform undefinedArg 1`] = `
"function undefinedArg(a, b) {
  while (true) {
    if (a < b) return a;
    a = a - 1;
    b = undefined;
  }
}"
`;
